openeo/NDVI_Timeseries_UDF_Smoothing.ipynb

```diff
@@ -0,0 +1,296 @@
+      "text/plain": [
+       "<Figure size 640x480 with 2 Axes>"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "\n",
+    "\n",
+    "g = scipy.signal.windows.gaussian(11, std=1.6)\n",
+    "kernel = np.outer(g, g)\n",
+    "kernel = kernel / kernel.sum()\n",
+    "im = plt.imshow(kernel)\n",
+    "plt.colorbar(im);"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "\n",
+    "classification = cube.band(\"SCL\")\n",
+    "mask = ~ ((classification == 4) | (classification == 5))\n",
+    "mask = mask.apply_kernel(kernel)\n",
+    "mask = mask > 0.1\n",
+    "\n",
+    "masked_ndvi = ndvi.mask(mask)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "With the `DataCube.aggregate_spatial()` method , we can calculate the mean NDVI for each of the fields.\n",
+    "\n",
+    "We now use a UDF (user-defined function)\n",
+    "to interpolate the missing values \n",
+    "and to apply a Savitzky-Golay filter for smoothing of the timeseries, using `scipy.signal.savgol_filter`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "udf = openeo.UDF(\"\"\"\n",
+    "from scipy.signal import savgol_filter\n",
+    "from openeo.udf import XarrayDataCube\n",
+    "\n",
+    "def apply_datacube(cube: XarrayDataCube, context: dict) -> XarrayDataCube:\n",
+    "    array = cube.get_array()\n",
+    "    filled = array.interpolate_na(dim='t')\n",
+    "    smoothed_array = savgol_filter(filled.values, 5, 2, axis=0)\n",
+    "    return DataCube(xarray.DataArray(smoothed_array, dims=array. dims,coords=array.coords))\n",
+    "\"\"\")\n",
+    "\n",
+    "smoothed_ndvi = masked_ndvi.apply_dimension(code=udf, dimension=\"t\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now, aggregate this again per field and get the time series."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "smoothed_timeseries = smoothed_ndvi.aggregate_spatial(geometries=fields, reducer=\"mean\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "smoothed_timeseries.download(\"output/timeseries-smoothed.json\", format=\"JSON\")"
```

Comment:
I would drop the unnecessary `format=JSON`

Reply:
updated